### DISCUSSION TITLE
What is the difference between "make prod" and "make install"?

### DISCUSSION THREAD

**Discussion by bil-inagaki** [processone/ejabberd#4324] [Questions] [answered]
I noticed a marked difference in the size of the beam created by “make prod” and “make install”.
"make prod" is smaller than "make install".

What is the reason?
Is there any performance difference?

ejabberd version: 24.07

Thanks.

#ejabberd

**Comment by badlop** [comment, marked as answer]
> "make prod" is smaller than "make install". What is the reason?

`make install` was manually handcrafted many years ago, it simply copies the files that were compiled to the defined destination.

I say 'simply', but in reality our `Makefile.in` includes a lot of code to correctly create the destination paths, copy all the files to the desired destination, etc: [see section `copy-files`](https://github.com/processone/ejabberd/blob/7fce7d1049d5e733e4ab1c078214cd415dda6bd3/Makefile.in#L276) from line 279 to 362.

---

`make prod` calls [`rebar3 release`](https://rebar3.org/docs/deployment/releases/) or [`mix release`](https://hexdocs.pm/mix/1.16.0/Mix.Tasks.Release.html) and asks them to build the release. Those building tools use [OTP release libraries](https://www.erlang.org/doc/system/release_structure.html#content) to actually build the release (see also [Adopting Erlang: Releases](https://adoptingerlang.org/docs/production/releases/)).

Those tools that take care to prepare the release will also take care to [reduce the size of beam files](https://rebar3.org/docs/deployment/releases/#modes):

> The prod mode expands to options that are commonly used for production releases: ... debug_info is set to strip, which makes BEAM files smaller and only removes data that is used by tools you more than likely aren’t including in your release

---

Notice that [there's a new `make install-rel`](https://www.process-one.net/blog/ejabberd-24-02/#new-make-install-rel-and-make-uninstall-rel) which is based in `make prod`, so you can benefit from all those improvements when installing ejabberd. And that requires a lot less code in `Makefile.in`, see [section `copy-files-rel`](https://github.com/processone/ejabberd/blob/7fce7d1049d5e733e4ab1c078214cd415dda6bd3/Makefile.in#L365) from line 368 to 377

---

> Is there any performance difference?

There shouldn't be any difference: the beam files are loaded into memory at server start, which is done only once, and after that they are in memory, regardless if the included debug information or not, or what was their file size.

**Reply by bil-inagaki** [reply]
@badlop
Thanks for your answers.
I will check the articles you linked.
